java/src/main/java/com/anaconda/skein/credentials/CredentialProvider.java

```diff
@@ -0,0 +1,9 @@
+package com.anaconda.skein.credentials;
+
+import org.apache.hadoop.security.Credentials;
+import org.apache.hadoop.yarn.exceptions.YarnException;
+import java.io.IOException;
+
+public interface CredentialProvider {
```

Comment:
Rather than abstracting this out into an interface with two implementations, I'd make a single class with a method for adding each one, and a larger method for iterating through the spec and adding credentials for each provider. In python pseudocode:

```python
class CredentialManager:
    def add_credentials_common(self, credentials):
        pass

    def add_credentials_hive(self, credentials, uri, principal):
        pass

    def add_credentials(self, credentials, spec):
        self.add_credentials_common(credentials)
        for cred in spec.credential_providers:
            if cred.name == "hive":
                self.add_credentials_hive(credentials, cred.uri, cred.principal)
            elif cred.name == ...:
                pass  # other implementations added here
```

This may not be as java idiomatic, but better matches with the existing code style. I don't see a reason to abstract this out when we'll only ever support a set list of credential providers.

Reply:
True, I'll refactor it this way.